src/framework/core/js/RemoteModel.js

```diff
@@ -0,0 +1,201 @@
+            // component's model state that should be updated with the remote source.
+            local: {},
+            remote: {},
+            requestInFlight: false
+        },
+        modelListeners: {
+            "requestInFlight": {
+                listener: "fluid.remoteModelComponent.launchPendingRequest",
+                args: ["{that}"]
+            }
+        },
+        listeners: {
+            "afterFetch.updateModel": {
+                listener: "fluid.remoteModelComponent.updateModelFromFetch",
+                args: ["{that}", "{arguments}.0"],
+                priority: "before:unblock"
+            },
+            "afterFetch.unblock": {
+                listener: "fluid.remoteModelComponent.unblockFetchReq",
+                args: ["{that}"]
+            },
+            "onFetchError.unblock": {
+                listener: "fluid.remoteModelComponent.unblockFetchReq",
+                args: ["{that}"]
+            },
+            "afterWrite.unblock": {
+                changePath: "requestInFlight",
+                value: false
+            },
+            "onWriteError.unblock": {
+                changePath: "requestInFlight",
+                value: false
+            }
+        },
+        invokers: {
+            fetch: {
+                funcName: "fluid.remoteModelComponent.fetch",
+                args: ["{that}"]
+            },
+            fetchImpl: "fluid.notImplemented",
+            write: {
+                funcName: "fluid.remoteModelComponent.write",
+                args: ["{that}"]
+            },
+            writeImpl: "fluid.notImplemented"
+        }
+    });
+
+    fluid.remoteModelComponent.launchPendingRequest = function (that) {
+        if (!that.model.requestInFlight) {
+            if (that.pendingRequests.fetch) {
+                that.fetch();
+            } else if (that.pendingRequests.write) {
+                that.write();
+            }
+        }
+    };
+
+    fluid.remoteModelComponent.updateModelFromFetch = function (that, fetchedModel) {
+        var remoteChanges = fluid.modelPairToChanges(fetchedModel, that.model.remote, "local");
+        var localChanges = fluid.modelPairToChanges(that.model.local, that.model.remote, "local");
+        var changes = remoteChanges.concat(localChanges);
+
+        // perform model updates in a single transaction
+        var transaction = that.applier.initiate();
+        transaction.fireChangeRequest({path: "local", type: "DELETE"}); // clear old local model
+        transaction.change("local", that.model.remote); // reset local model to the base for applying changes.
+        transaction.fireChangeRequest({path: "remote", type: "DELETE"}); // clear old remote model
+        transaction.change("remote", fetchedModel); // update remote model to fetched changes.
+        fluid.fireChanges(transaction, changes); // apply changes from remote and local onto base model.
```

Comment:
I don't know about you, but to me something seems interestingly and fundamentally right about this (even though there may well still be bugs in detail)